nservicebus/architecture/publishing-from-web-applications.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Publishing Events from Web Applications
+summary: "Guidance on why you should or should not publish events from a web application"
+tags: []
+redirects:
+---
+
+You may have heard that [it's bad practice to publish events from a web application](http://www.make-awesome.com/2010/10/why-not-publish-nservicebus-messages-from-a-web-application/). This guidance is a bit outdated, as improvements Microsoft has made to IIS, more varied methods of hosting .NET web applications, and the addition of additional NServiceBus transports have made things much more nuanced.
+
+Yes, it is possible, and in some circumstances, even advisable to publish events from within a web application. This article will describe the guidelines for publishing messages within web applications under different circumstances.
+
+## Why/Why Not?
+
+Historically, there have been good reasons to avoid publishing events from within web applications:
+
+1. **Transactions and Consistency** - Generally good events should announce that something has *already* happened. Many times this involves making some changes to a database and then publishing the result. But HTTP is inherently unreliable and does not have built-in retries. If an exception occurs before the event is published, you may lose your only opportunity to publish that event. In these circumstances, it's generally better to send a command with the payload of the HTTP request, and have another endpoint process that command with the advantages of automatic retry and an error queue.
+2. **Web Application Scale-Out** - Because events can only be published from one logical publisher, it can be problematic to scale out a web application that is publishing events. More on this later.
+3. **Web App Recycling** - Prior to IIS 7.5 introducing the AlwaysRunning feature (AKA AlwaysOn), web applications not receiving could be recycled or idled at any time, which would result in incoming messages building up in an input queue, without an application actively processing them.

```

Comment:
"web applications not receiving" doesn't read well for me. Change to "Idle web applications could be shutdown or recycled at any time" ?

Do we need to split shutdown vs recycled since they are associated with different problems:

Shutdown: Messages pile up since there is no receiving process running
Recycle: You might "miss" message functionally to be triggered of you have callbacks or publishing over signalr. Is recycle even relevant for the "publishes" ? (I'd say only if the publish is over signalr)

Actually is the relevant at all? Or should this article be a more generic "Dangers when receiving from a webapp"?

Or is that a new different guidance page?